COPYRIGHT

```diff
@@ -23,66 +23,55 @@ The Rust Project includes packages written by third parties.
-          and/or other materials provided with the
-          distribution.
+    Copyright (C) 2008-2010 Tachyon Technologies.
+    All rights reserved.
 
-        * Neither the names of the LLVM Team, University of
-          Illinois at Urbana-Champaign, nor the names of its
-          contributors may be used to endorse or promote
-          products derived from this Software without
-          specific prior written permission.
+    Redistribution and use in source and binary forms, with
+    or without modification, are permitted provided that the
+    following conditions are met:
 
-    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
-    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
-    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
-    PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT
-    SHALL THE CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE
-    FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
-    ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
-    OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
-    OTHER DEALINGS WITH THE SOFTWARE.
+    1. Redistributions of source code must retain the above
+       copyright notice, this list of conditions and the
+       following disclaimer.
 
-* Additional libraries included in LLVM carry separate
-  BSD-compatible licenses. See src/llvm-project/llvm/LICENSE.TXT
-  for details.
+    2. Redistributions in binary form must reproduce the
+       above copyright notice, this list of conditions and
+       the following disclaimer in the documentation and/or
+       other materials provided with the distribution.
 
-* compiler-rt, in src/compiler-rt is dual licensed under
-  LLVM's license and MIT:
+    THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR
+    IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    DEVELOPERS AND CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+    INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+    PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+    USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+    NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+    USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+    OF SUCH DAMAGE.
 
-    Copyright (c) 2009-2014 by the contributors listed in
-    CREDITS.TXT
+* The rustc_apfloat crate, located in compiler/rustc_apfloat, is based
+  on the APFloat library from a version of LLVM licensed under the UIUC
+  license. Contributions to the rustc_apfloat crate are licensed either
+  under the standard Rust licensing (Apache License, Version 2.0 or the
+  MIT license, at your option) or under the Apache License, Version 2.0,
```

Comment:
I think this is not quite right -- Rust's standard licensing requires that contributions are licensed under *both* Apache and MIT, not one of them at the contributor's option, right? Otherwise *users* cannot choose one or the other, if there's code under just one of them, presumably?

(But maybe I have misunderstood).

Reply:
So, I copied the wording from the top of this file: 

https://github.com/rust-lang/rust/blob/c0672870491e84362f76ddecd50fa229f9b06dff/COPYRIGHT#L16-L19

I believe the wording here is correct as we want contributions to rustc_apfloat to be contributed under all of Apache, MIT, UIUC and to then allow the user to decide what license they will comply with.

Reply:
I think the problem is the "Contributions to the" prefix -- that is, the code is licensed under (Apache 2.0 or MIT or UIUC), that's correct, but *contributions* must be licensed under that same set of licenses -- contributors can't choose to only license their changes under MIT, for example.

Reply:
Agreed! I'm just struggling to see what other word would be appropriate here. The issue is that I'm specifically trying to make a distinction between the initial port and subsequent changes to that code. I can say "Subsequent changes to the library ..." but to my mind that means the same thing so it doesn't solve the issue. 

I'm interpreting "you" in this document to be a downstream user of Rust, not a contributor. As such, "at your option" is referring to options the user has not options a contributor has. Perhaps that's not the correct interpretation?

Reply:
Yeah. I think it makes sense from you = user perspective, but then it's not entirely clear to me what users are expected to do with the 'based on' language. It also seems like we're needlessly(?) repeating ourselves -- MIT and Apache 2.0 are both listed twice and seemingly under the same "or" grouping from user's perspective, right?

Maybe we can change to something like:

> The rustc_apfloat crate, located in compiler/rustc_apfloat, is based on the APFloat library from a version of LLVM licensed under the UIUC license. Subsequent contributions to the rustc_apfloat crate are licensed under the Apache License, Version 2.0, the MIT license, or the UIUC license, at your option.

I think the subsequent to me helps a little, as does the simplification away from having repeated licenses -- it makes the sentence structure simpler and overall gets the same message across, I believe.

I think the takeaway here is probably that it's probably worth doing some work on providing a "contributor copyright" file of some kind that basically requests licensing from contributors under some set of licenses. If possible, I think ideally we'd structure things so that we could omit UIUC from the list here, so that the story for contributors to current rust-lang/rust would be consistently "license all changes under both Apache 2.0 and MIT (each separately)", which seems better than having some folders also require other licenses to be 'granted'.

Reply:
> I think the subsequent to me helps a little, as does the simplification away from having repeated licenses -- it makes the sentence structure simpler and overall gets the same message across, I believe.

Agreed, it is much simpler! I believe the problem though is that until now, we've not asked that changes to this code be triple licensed and the advice Felix and I were given said this likely means those contributions were done under the normal Rust license (ie, only MIT/Apache2 only). Going forward, we're going to require contributors to `rustc_apfloat` to triple license under MIT/Apache2/UIUC but that still leaves the contributions which have already been made. As such, I don't think saying "Subsequent contributions to the `rustc_apfloat` crate are licensed under the Apache License, Version 2.0, the MIT license, or the UIUC license, at your option." is entirely correct. 

We could certainly try to go back and get permission from contributors to license their code this way and that would simplify things here. I didn't want to block these improvements on that happening and it's possible we may not get every contributor to sign off on that. 

Per your other comment, I'm very much in agreement we should verify the specific changes with the Foundation but I'd like for us to decide what the changes should look prior to engaging with their lawyers again. 

Looking to the future, I see two potential paths to simplify our licensing here:

1. If we can secure relicensing consent from past contributors, we could simplify the licensing here to simply say "The `rustc_apfloat` crate is licensed under the UIUC license". That isn't ideal as it still deviates from our standard licensing but at least it is easy to explain and understand. 

2. We can perform a clean-room re-implementation of the `rustc_apfloat` crate based only on the type & function signatures of the current code. The code is quite self-contained and these should be quite feasible but Felix and I are not aware of anyone interested in taking this on. We'd also require some additional collaboration with the RF to make sure the re-implementation is done correctly and in a well documented way so there is no question as to the provenance of the code.